packages/@lwc/engine/src/framework/restrictions.ts

```diff
@@ -382,12 +388,49 @@ function getComponentRestrictionsDescriptors(): PropertyDescriptorMap {
     };
 }
 
-function getLightningElementPrototypeRestrictionsDescriptors(proto: object): PropertyDescriptorMap {
+function getLightningElementPrototypeRestrictionsDescriptors(
+    proto: typeof LightningElement.prototype
+): PropertyDescriptorMap {
     if (process.env.NODE_ENV === 'production') {
         // this method should never leak to prod
         throw new ReferenceError();
     }
-    const descriptors = {};
+
+    const originalDispatchEvent = proto.dispatchEvent;
+
+    const descriptors = {
+        dispatchEvent: generateDataDescriptor({
+            value(this: LightningElement, event: Event): boolean {
+                const vm = getAssociatedVM(this);
+
+                assert.isFalse(
+                    isBeingConstructed(vm),
+                    `this.dispatchEvent() should not be called during the construction of the custom element for ${getComponentAsString(
+                        this
+                    )} because no one is listening just yet.`
+                );
+
+                if (!isNull(event) && isObject(event)) {
+                    const { type } = event;
+
+                    if (isString(type) && !/^[a-z][a-z0-9_]*$/.test(type)) {
```

Comment:
I think this condition is incorrect, type should be a string, and a valid one, otherwise throw.